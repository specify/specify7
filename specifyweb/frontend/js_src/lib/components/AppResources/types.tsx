import type { LocalizedString } from 'typesafe-i18n';

import { preferencesText } from '../../localization/preferences';
import { reportsText } from '../../localization/report';
import { resourcesText } from '../../localization/resources';
import type { IR, RA, RR } from '../../utils/types';
import { ensure } from '../../utils/types';
import { icons } from '../Atoms/Icons';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpAppResourceDir, Tables } from '../DataModel/types';
import type { AppResourceMode } from './helpers';

export type AppResourceScope =
  | 'collection'
  | 'discipline'
  | 'global'
  | 'user'
  | 'userType';

export type AppResourceType = {
  readonly tableName: keyof Tables & ('SpAppResource' | 'SpViewSetObj');
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
};

export const appResourceTypes: RR<AppResourceMode, AppResourceType> = {
  appResources: {
    tableName: 'SpAppResource',
    icon: icons.cog,
    label: resourcesText.appResource(),
  },
  viewSets: {
    tableName: 'SpViewSetObj',
    icon: icons.pencilAt,
    label: resourcesText.formDefinitions(),
  },
};

export type ScopedAppResourceDir = SerializedResource<SpAppResourceDir> & {
  readonly scope: AppResourceScope;
};

type AppResourceSubType = {
  readonly mimeType: string | undefined;
  readonly name: string | undefined;
  readonly documentationUrl: string | undefined;
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
  /**
   * Whether when creating a new app resource of this type, should copy the
   * contents from an existing app resource of that type that is in current
   * scope.
   * Default value:
   * If app resource type can only have one specific name, this is true
   * Else false
   */
  readonly useTemplate?: boolean;
  /**
   * Only allow creating this app resource at certain levels
   */
  readonly scope?: RA<AppResourceScope>;
};

/**
 * The order of the subtypes matters. The filtering algorithm loops over these
 * in the order they are defined to find the first subType that matches the
 * current resource. Thus, subtypes should be sorted from the most
 * specific to the least specific.
 */
export const appResourceSubTypes = ensure<IR<AppResourceSubType>>()({
  label: {
    mimeType: 'jrxml/label',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.ticket,
    label: reportsText.label(),
  },
  report: {
    mimeType: 'jrxml/report',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.documentReport,
    label: reportsText.report(),
  },
  userPreferences: {
    mimeType: 'application/json',
    name: 'UserPreferences',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/specify-7-user-preferences-webinar/861',
    icon: icons.cog,
    label: preferencesText.userPreferences(),
    useTemplate: false,
    scope: ['user'],
  },
  defaultUserPreferences: {
    mimeType: 'application/json',
    name: 'DefaultUserPreferences',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Setting-default-user-preferences',
    icon: icons.cog,
    label: preferencesText.defaultUserPreferences(),
  },
  collectionPreferences: {
    mimeType: 'application/json',
    name: 'CollectionPreferences',
    documentationUrl: undefined,
    icon: icons.cog,
    label: preferencesText.collectionPreferences(),
    scope: ['collection'],
  },
  leafletLayers: {
    mimeType: 'application/json',
    name: 'leaflet-layers',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Adding-Custom-Tile-Servers',
    icon: icons.locationMarker,
    label: resourcesText.leafletLayers(),
  },
  rssExportFeed: {
    mimeType: 'text/xml',
    name: 'ExportFeed',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Darwin-Core-Archive-Publishing',
    icon: icons.upload,
    label: resourcesText.rssExportFeed(),
  },
  expressSearchConfig: {
    mimeType: 'text/xml',
    name: 'ExpressSearchConfig',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/simple-search-config/183',
    icon: icons.search,
    label: resourcesText.expressSearchConfig(),
  },
  typeSearches: {
    mimeType: 'text/xml',
    name: 'TypeSearches',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/adding-a-non-native-query-combo-box/859#h-1-type-search-definition-typesearch_defxml-8',
    icon: icons.documentSearch,
    label: resourcesText.typeSearches(),
  },
  webLinks: {
    mimeType: 'text/xml',
    name: 'WebLinks',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/weblinks.xml',
    icon: icons.externalLink,
    label: resourcesText.webLinks(),
  },
  uiFormatters: {
    mimeType: 'text/xml',
    name: 'UIFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/uiformatters.xml',
    icon: icons.hashtag,
    label: resourcesText.uiFormatters(),
  },
  dataObjectFormatters: {
    mimeType: 'text/xml',
    name: 'DataObjFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/dataobj_formatters.xml',
    icon: icons.variable,
    label: resourcesText.dataObjectFormatters(),
  },
  dataEntryTables: {
    mimeType: 'text/xml',
    name: 'DataEntryTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/fish/dataentry_task.xml',
    icon: icons.pencilAt,
    label: resourcesText.dataEntryTables(),
  },
  interactionsTables: {
    mimeType: 'text/xml',
    name: 'InteractionsTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/interactionstask.xml',
    icon: icons.chat,
    label: resourcesText.interactionsTables(),
  },
  otherXmlResource: {
    mimeType: 'text/xml',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.code,
    label: resourcesText.otherXmlResource(),
  },
  otherJsonResource: {
    mimeType: 'application/json',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.adjustments,
    label: resourcesText.otherJsonResource(),
  },
  otherPropertiesResource: {
    mimeType: 'text/x-java-properties',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.viewList,
    label: resourcesText.otherPropertiesResource(),
  },
  otherAppResources: {
    mimeType: undefined,
    name: undefined,
    documentationUrl: undefined,
    icon: icons.document,
    label: resourcesText.otherAppResource(),
  },
} as const);
