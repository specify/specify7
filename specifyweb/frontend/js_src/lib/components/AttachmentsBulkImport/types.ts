import type React from 'react';
import type { LocalizedString } from 'typesafe-i18n';
import type { State } from 'typesafe-reducer';

import type { RA, RR } from '../../utils/types';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { Attachment, Tables } from '../DataModel/types';
import type { DatasetBase, DatasetBriefBase } from '../WbPlanView/Wrapped';
import type { PartialAttachmentUploadSpec } from './Import';
import type { staticAttachmentImportPaths } from './importPaths';
import type { keyLocalizationMapAttachment } from './utils';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentLocation: string;
};

type Matched = State<
  'matched',
  {
    readonly id: number;
  }
>;

type Skipped = State<
  'cancelled' | 'skipped',
  {
    readonly reason: keyof typeof keyLocalizationMapAttachment;
  }
>;
type Success = State<
  'success',
  {
    readonly successType: 'deleted' | 'uploaded';
  }
>;
export type AttachmentStatus = Matched | Skipped | Success;

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'uploadFile'>;

type UploadableFileSpec = {
  readonly uploadFile: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
  /*
   * This is added because if an attachment was uploaded correctly, but a network error
   * happened while saving the resource, and we retry, we will loose the previous attachment
   */
  readonly attachmentFromPreviousTry?: SpecifyResource<Attachment>;
};

/*
 * Forcing keys to be primitive because objects would be
 * ignored during syncing to backend.
 */
export type BoundFile = Pick<File, 'name' | 'size' | 'type'>;

export type UnBoundFile = {
  readonly file: BoundFile | File;
  readonly parsedName?: string;
};

export type AttachmentWorkProgress = {
  readonly total: number;
  readonly uploaded: number;
  readonly type: 'interrupted' | 'safe' | 'stopped' | 'stopping';
  readonly retryingIn: number;
  readonly stoppedByUser: boolean;
};

export type AttachmentWorkRef = {
  // eslint-disable-next-line functional/prefer-readonly-type
  mappedFiles: RA<PartialUploadableFileSpec>;
  // eslint-disable-next-line functional/prefer-readonly-type
  uploadPromise: Promise<number | undefined>;
  readonly retrySpec: Record<number, number>;
};

export type AttachmentWorkStateProps = {
  readonly onCompletedWork: (
    uploadables: RA<PartialUploadableFileSpec> | undefined
  ) => void;
  readonly workProgress: AttachmentWorkProgress;
  readonly workRef: React.MutableRefObject<AttachmentWorkRef>;
  readonly onStop: () => void;
  readonly triggerNow: () => void;
  readonly dialogText: RR<
    'onAction' | 'onCancelled' | 'onCancelledDescription',
    LocalizedString
  >;
};

export type AttachmentDatasetBrief = DatasetBriefBase & {
  readonly uploaderstatus:
    | 'deleting'
    | 'deletingInterrupted'
    | 'main'
    | 'uploading'
    | 'uploadInterrupted'
    | 'validating';
};

export type AttachmentDataSetPlan = AttachmentDatasetBrief & {
  readonly uploadplan: {
    readonly staticPathKey: string;
  };
};

export type AttachmentDataSet = AttachmentDatasetBrief &
  DatasetBase & {
    readonly rows: RA<PartialUploadableFileSpec>;
    readonly uploadplan: PartialAttachmentUploadSpec;
    readonly uploadresult?: {
      readonly timestamp: string;
    };
  };

export type FetchedDataSet =
  | AttachmentDataSet &
      (
        | { readonly uploaderstatus: 'main' }
        | ({
            readonly uploaderstatus: 'deleting' | 'uploading';
            readonly rows: RA<PartialUploadableFileSpec>;
          } & {
            readonly uploadplan: {
              readonly staticPathKey: keyof typeof staticAttachmentImportPaths;
            };
          })
      );

export type WrappedActionProps<KEY extends keyof Tables> = {
  readonly uploadableFile: PartialUploadableFileSpec;
  readonly baseTableName: KEY;
  readonly dryRun: boolean;
  readonly triggerRetry?: () => void;
};
