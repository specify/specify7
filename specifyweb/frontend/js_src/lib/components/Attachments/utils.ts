import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
