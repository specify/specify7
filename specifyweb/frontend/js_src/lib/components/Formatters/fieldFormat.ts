import type { LocalizedString } from 'typesafe-i18n';

import type { Parser } from '../../utils/parser/definitions';
import { resolveParser } from '../../utils/parser/definitions';
import { parseValue } from '../../utils/parser/parse';
import { localized } from '../../utils/types';
import { removeKey } from '../../utils/utils';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { LiteralField } from '../DataModel/specifyField';
import type { PickList } from '../DataModel/types';
import { unsafeGetPickLists } from '../PickLists/definitions';
import { fetchPickList, getPickListItems } from '../PickLists/fetch';

/*
 * BUG: when formatting a date field, it uses the databaseDateFormat rather
 *  than fullDateFormat(). check usages to see where this may be a problem.
 */

/**
 * Format value for output
 *
 * @remarks
 * Parses the value
 * Formats it
 * Runs UI formatter if needed
 * Finds pickList item if available
 */
export async function fieldFormat(
  field: LiteralField | undefined,
  value: boolean | number | string | null | undefined,
  parser?: Parser,
  formatter?: string
): Promise<LocalizedString> {
  if (value === undefined || value === null) return localized('');

  const formatted = uiFormatter(field, value?.toString() ?? '', formatter);
  if (typeof formatted === 'string') return formatted;

  // Find Pick List Item title
  const pickListName = parser?.pickListName ?? field?.getPickList();
  if (typeof pickListName === 'string') {
    const pickList = await fetchPickList(pickListName);
    const formatted = formatPickList(pickList, value);
    if (typeof formatted === 'string') return formatted;
  }

  return formatValue(field, parser, value, false);
}

function uiFormatter(
  field: Partial<LiteralField> | undefined,
  value: string,
  formatter?: string
): LocalizedString | undefined {
  const uiFormatter = field?.getUiFormatter?.();
  if (typeof uiFormatter === 'object') {
    const formatted = uiFormatter.format(value?.toString() ?? '');
    if (typeof formatted === 'string') return formatted;
    else
      console.error(
        `Invalid value for ${
          formatter ?? field?.getFormat?.() ?? ''
        } formatter: ${value}`,
        { field }
      );
  }
  return undefined;
}

function formatPickList(
  pickList: SpecifyResource<PickList> | undefined,
  value: boolean | number | string
): LocalizedString | undefined {
  if (pickList === undefined) return undefined;
  const parsedValue = value.toString();
  const items = getPickListItems(pickList);
  const item = items.find((item) => item.value === parsedValue);
  return localized(item?.title);
}

/**
 * Format fields value. Does not format pick list items.
 * Prefer using fieldFormat() or syncFieldFormat() instead
 */
function formatValue<STRICT extends boolean = false>(
  field: LiteralField | undefined,
  parser: Parser | undefined,
  value: boolean | number | string,
  // @ts-expect-error
  strict: STRICT = false
): STRICT extends true ? LocalizedString | undefined : LocalizedString {
  const resolvedParser = parser ?? resolveParser(field ?? {});

  const parseResults = parseValue(
    removeKey(resolvedParser, 'required'),
    undefined,
    value.toString()
  );
  if (parseResults.isValid)
    return localized(
      resolvedParser.printFormatter?.(parseResults.parsed, resolvedParser) ??
        (parseResults.parsed as string | undefined)?.toString() ??
        ''
    );
  else
    console.error('Failed to parse value for field', {
      field,
      resolvedParser,
      parseResults,
    });

  return (strict ? undefined : localized(value.toString())) as ReturnType<
    typeof formatValue<STRICT>
  >;
}

/**
 * Like fieldFormat, but synchronous, because it doesn't fetch a pick list if
 * it is not already fetched
 */

export function syncFieldFormat<STRICT extends boolean = false>(
  field: LiteralField | undefined,
  value: boolean | number | string | null | undefined,
  parser?: Parser,
  formatter?: string,
  // @ts-expect-error
  strict: STRICT = false
): ReturnType<typeof formatValue<STRICT>> {
  if (value === undefined || value === null) return '' as LocalizedString;

  const formatted = uiFormatter(field, value?.toString() ?? '', formatter);

  if (typeof formatted === 'string') return formatted;

  // Find Pick List Item Title
  const pickListName = parser?.pickListName ?? field?.getPickList();
  if (typeof pickListName === 'string') {
    const pickList = unsafeGetPickLists()[pickListName];
    const formatted = formatPickList(pickList, value);
    if (typeof formatted === 'string') return formatted;
  }
  return formatValue<STRICT>(field, parser, value, strict);
}

export const exportsForTests = {
  formatValue,
  formatPickList,
};
