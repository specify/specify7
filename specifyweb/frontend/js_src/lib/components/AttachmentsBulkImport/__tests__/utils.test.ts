import { requireContext } from '../../../tests/helpers';
import { formatterToParser } from '../../../utils/parser/definitions';
import type { IR, RA } from '../../../utils/types';
import { localized } from '../../../utils/types';
import { tables } from '../../DataModel/tables';
import {
  CatalogNumberNumeric,
  formatterTypeMapper,
  UiFormatter,
} from '../../FieldFormatters';
import { syncFieldFormat } from '../../Formatters/fieldFormat';
import type { PartialUploadableFileSpec, UnBoundFile } from '../types';
import {
  inferDeletedAttachments,
  inferUploadedAttachments,
  matchFileSpec,
  resolveFileNames,
} from '../utils';

requireContext();

type TestDefinition = IR<{
  readonly uiFormatter: UiFormatter;
  readonly testCases: RA<readonly [string, string | undefined]>;
}>;

const staticTestCases: TestDefinition[string]['testCases'] = [
  ['      .jpg', undefined],
  ['\n\n\n\n\t\t', undefined],
  ['\t\t     \n\n.jpg', undefined],
];

const fileNameTestSpec: TestDefinition = {
  catalogNumber: {
    uiFormatter: new CatalogNumberNumeric(),
    testCases: [
      ['000000001.jpg', '000000001'],
      ['00002.jpg', '000000002'],
      ['00001(1).jpg', undefined],
      ['000000003(2).jpg', '000000003'],
      // BUG: This could lead to unexpected matches
      ['0000000000002.jpg', '000000000'],
      ['someRandomValue.jpg', undefined],
      ['\t\t\t0000 \n\n.jpg', '000000000'],
    ],
  },
  numeric: {
    uiFormatter: new UiFormatter(
      false,
      localized('testNumeric'),
      [
        new formatterTypeMapper.numeric({
          size: 3,
          autoIncrement: true,
          byYear: false,
        }),
      ],
      tables.CollectionObject
    ),
    testCases: [
      ['000.jpg', '000'],
      ['001.jpg', '001'],
      ['23.jpg', '023'],
      ['DEF001.jpg', undefined],
      ['000(1).jpg', '000'],
      ['23(1).jpg', undefined],
      ['\t\t\t02 \n\n.jpg', '002'],
    ],
  },
  regex: {
    uiFormatter: new UiFormatter(
      false,
      localized('testRegex'),
      [
        new formatterTypeMapper.regex({
          size: 3,
          autoIncrement: true,
          value: localized('^\\d{1,6}(?:[a-zA-Z]{1,2})?$'),
          byYear: false,
        }),
      ],
      tables.CollectionObject
    ),
    testCases: [
      ['45265.jpg', '45265'],
      ['45622AB.jpg', '45622AB'],
      ['45622AB', '45622AB'],
      ['45622787ABCDEF.jpg', undefined],
    ],
  },
};

describe('file names resolution test', () => {
  // Using text1 to not confuse with potential catalog number autonumbering
  Object.entries(fileNameTestSpec).forEach(
    ([testName, { uiFormatter, testCases }]) => {
      const allTestCases = [...testCases, ...staticTestCases];
      test(testName, () => {
        jest.spyOn(console, 'error').mockImplementation();
        const field = tables.CollectionObject.getLiteralField('text1')!;
        const getResultFormatter =
          (formatter: UiFormatter) => (value: number | string | undefined) =>
            value === undefined || value === null
              ? undefined
              : syncFieldFormat(
                  field,
                  value.toString(),
                  formatterToParser(field, formatter),
                  undefined,
                  true
                );
        const resultFormatter = getResultFormatter(uiFormatter);
        allTestCases.forEach(([input, output]) => {
          expect(resolveFileNames(input, resultFormatter, uiFormatter)).toEqual(
            output
          );
        });
      });
    }
  );
});

describe('resolve file names', () => {
  const queryResults = [
    [0, '1'] as const,
    [1, '2'] as const,
    [2, '3'] as const,
    [3, '4'] as const,
    [10, '5'] as const,
    [11, '5'] as const,
  ];
  const uploadSpec: RA<PartialUploadableFileSpec> = [
    {
      uploadFile: {
        file: { name: 'test', size: 0, type: '0' },
        parsedName: '1',
      },
    },
    {
      uploadFile: {
        file: { name: 'test2', size: 0, type: '0' },
        parsedName: '2',
      },
    },
    {
      uploadFile: {
        file: { name: 'test3', size: 0, type: '0' },
        parsedName: '3',
      },
    },
    {
      uploadFile: {
        file: { name: 'test4', size: 0, type: '0' },
        parsedName: '4',
      },
      attachmentId: 10,
      matchedId: [2],
      disambiguated: 10,
    },
    {
      /*
       * Shouldn't assume that backend will give back all the results back
       * just more than once, since MySQL will remove duplicates when having the in (..)
       * clause
       */
      uploadFile: {
        file: { name: 'test5', size: 0, type: '0' },
        parsedName: '4',
      },
    },
    {
      uploadFile: {
        file: { name: 'test6', size: 0, type: '0' },
        parsedName: '5',
      },
      // This will be reset
      matchedId: [12, 13, 10, 11],
      disambiguated: 10,
    },
    {
      uploadFile: {
        file: { name: 'test6', size: 0, type: '0' },
        parsedName: '5',
      },
      matchedId: [10, 11],
      // This should be reset, even if keeping disambiguation since it is stale
      disambiguated: 18,
    },
    {
      uploadFile: {
        file: { name: 'test7', size: 0, type: '0' },
        parsedName: '5',
      },
      matchedId: [10, 11],
      // This should be preserved
      disambiguated: 10,
    },
  ];
  test('resolve file names from validation', () => {
    expect(matchFileSpec(uploadSpec, queryResults)).toMatchSnapshot();
  });
  test('resolve file names from validation (keeping disambiguation)', () => {
    expect(matchFileSpec(uploadSpec, queryResults, true)).toMatchSnapshot();
  });
});

const fakeFile: UnBoundFile = {
  file: {
    name: 'someName',
    size: 0,
    type: 'test',
  },
};

test('reconstruct uploading attachment spec', () => {
  const queryResults = [
    [0, [1, 'location1.jpg']],
    [0, [2, 'location2.jpg']],
    [0, [3, 'location3.jpg']],
    [1, [4, 'location4.jpg']],
    [2, [10, null]], // If set to null by user, skip
  ] as RA<readonly [number, RA<number | string | null>]>;

  const files: RA<PartialUploadableFileSpec> = [
    {
      status: {
        type: 'matched',
        id: 0,
      },
      uploadTokenSpec: {
        attachmentLocation: 'location1.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
    {
      status: {
        type: 'matched',
        id: 0,
      },
      uploadTokenSpec: {
        attachmentLocation: 'location2.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
    {
      status: {
        type: 'matched',
        id: 0,
      },
      uploadTokenSpec: {
        attachmentLocation: 'location3.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
    // This file will be skipped in checking for attachment locations
    {
      status: {
        type: 'skipped',
        reason: 'noMatch',
      },
      uploadTokenSpec: {
        attachmentLocation: 'location10.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
    {
      status: {
        type: 'matched',
        // This resource's upload was interrupted.
        id: 100,
      },
      uploadTokenSpec: {
        attachmentLocation: 'location10.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
    {
      status: {
        type: 'matched',
        // This resource was uploaded, but then modified by the user in someway
        id: 2,
      },
      uploadTokenSpec: {
        attachmentLocation: 'locationAny.jpg',
        token: 'fakeToken',
      },
      uploadFile: fakeFile,
    },
  ];
  expect(inferUploadedAttachments(queryResults, files)).toMatchSnapshot();
});

test('reconstruct deleting attachment spec', () => {
  const queryResults = [
    [0, [1]],
    [1, [2]],
  ] as RA<readonly [number, RA<number>]>;
  const files: RA<PartialUploadableFileSpec> = [
    // This file wasn't deleted
    {
      status: {
        type: 'matched',
        id: 0,
      },
      uploadFile: fakeFile,
      attachmentId: 1,
    },
    {
      status: {
        type: 'matched',
        id: 0,
      },
      uploadFile: fakeFile,
      // This file was successfully deleted
      attachmentId: 100,
    },
    {
      status: {
        type: 'skipped',
        reason: 'incorrectFormatter',
      },
      uploadFile: fakeFile,
    },
    // This file was uploaded, but the corresponding resource was deleted,
    {
      status: {
        type: 'matched',
        id: 10,
      },
      uploadFile: fakeFile,
      attachmentId: 100,
    },
    // This file was also not deleted
    {
      status: {
        type: 'matched',
        id: 1,
      },
      uploadFile: fakeFile,
      attachmentId: 2,
    },
  ];
  expect(inferDeletedAttachments(queryResults, files)).toMatchSnapshot();
});
