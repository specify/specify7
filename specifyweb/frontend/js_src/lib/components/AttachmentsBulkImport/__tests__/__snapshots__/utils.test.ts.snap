// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reconstruct deleting attachment spec 1`] = `
[
  {
    "attachmentId": 1,
    "status": {
      "reason": "rollbackInterruption",
      "type": "cancelled",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
  },
  {
    "attachmentId": undefined,
    "status": {
      "successType": "deleted",
      "type": "success",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
  },
  {
    "status": {
      "reason": "incorrectFormatter",
      "type": "skipped",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
  },
  {
    "attachmentId": undefined,
    "status": {
      "successType": "deleted",
      "type": "success",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
  },
  {
    "attachmentId": 2,
    "status": {
      "reason": "rollbackInterruption",
      "type": "cancelled",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
  },
]
`;

exports[`reconstruct uploading attachment spec 1`] = `
[
  {
    "attachmentId": 1,
    "status": {
      "successType": "uploaded",
      "type": "success",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "location1.jpg",
      "token": "fakeToken",
    },
  },
  {
    "attachmentId": 2,
    "status": {
      "successType": "uploaded",
      "type": "success",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "location2.jpg",
      "token": "fakeToken",
    },
  },
  {
    "attachmentId": 3,
    "status": {
      "successType": "uploaded",
      "type": "success",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "location3.jpg",
      "token": "fakeToken",
    },
  },
  {
    "status": {
      "reason": "noMatch",
      "type": "skipped",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "location10.jpg",
      "token": "fakeToken",
    },
  },
  {
    "attachmentId": undefined,
    "status": {
      "reason": "uploadInterruption",
      "type": "cancelled",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "location10.jpg",
      "token": "fakeToken",
    },
  },
  {
    "attachmentId": undefined,
    "status": {
      "reason": "uploadInterruption",
      "type": "cancelled",
    },
    "uploadFile": {
      "file": {
        "name": "someName",
        "size": 0,
        "type": "test",
      },
    },
    "uploadTokenSpec": {
      "attachmentLocation": "locationAny.jpg",
      "token": "fakeToken",
    },
  },
]
`;

exports[`resolve file names resolve file names from validation (keeping disambiguation) 1`] = `
[
  {
    "disambiguated": undefined,
    "matchedId": [
      0,
    ],
    "uploadFile": {
      "file": {
        "name": "test",
        "size": 0,
        "type": "0",
      },
      "parsedName": "1",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      1,
    ],
    "uploadFile": {
      "file": {
        "name": "test2",
        "size": 0,
        "type": "0",
      },
      "parsedName": "2",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test3",
        "size": 0,
        "type": "0",
      },
      "parsedName": "3",
    },
  },
  {
    "attachmentId": 10,
    "disambiguated": 10,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test4",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      3,
    ],
    "uploadFile": {
      "file": {
        "name": "test5",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": 10,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test7",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
]
`;

exports[`resolve file names resolve file names from validation 1`] = `
[
  {
    "disambiguated": undefined,
    "matchedId": [
      0,
    ],
    "uploadFile": {
      "file": {
        "name": "test",
        "size": 0,
        "type": "0",
      },
      "parsedName": "1",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      1,
    ],
    "uploadFile": {
      "file": {
        "name": "test2",
        "size": 0,
        "type": "0",
      },
      "parsedName": "2",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test3",
        "size": 0,
        "type": "0",
      },
      "parsedName": "3",
    },
  },
  {
    "attachmentId": 10,
    "disambiguated": 10,
    "matchedId": [
      2,
    ],
    "uploadFile": {
      "file": {
        "name": "test4",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      3,
    ],
    "uploadFile": {
      "file": {
        "name": "test5",
        "size": 0,
        "type": "0",
      },
      "parsedName": "4",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test6",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
  {
    "disambiguated": undefined,
    "matchedId": [
      10,
      11,
    ],
    "uploadFile": {
      "file": {
        "name": "test7",
        "size": 0,
        "type": "0",
      },
      "parsedName": "5",
    },
  },
]
`;
